### PR TITLE
Return bad request for invalid stat parameters in stats API

### PR DESCRIPTION
**Pull request by q-andy**
### Description
Fixes bug in parsing stat names while calling the stats API. Currently it will log every stat specified when API is called. We change this to return a 400 bad request instead.

Technically this could be a breaking change, but only for users that are purposefully calling the API with invalid stat names, which doesn't have any valid use case.

Example 400:
```
GET /_plugins/_neural/stats/text_embedding
{
	"error": {
		"root_cause": [
			{
				"type": "illegal_argument_exception",
				"reason": "request [/_plugins/_neural/stats/text_embedding] contains unrecognized stat: [text_embedding] -> did you mean [text_embedding_executions]?"
			}
		],
		"type": "illegal_argument_exception",
		"reason": "request [/_plugins/_neural/stats/text_embedding] contains unrecognized stat: [text_embedding] -> did you mean [text_embedding_executions]?"
	},
	"status": 400
}
```
### Related Issues
n/a

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/neural-search/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by weijia-aws — APPROVED** [review]

File: src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsAction.java [hunk 1]
```diff
@@ -167,7 +168,7 @@ private NeuralStatsInput createNeuralStatsInputFromRequestParams(RestRequest req
         }
 
         // Process requested stats
-        boolean anyStatAdded = processRequestedStats(stats.get(), neuralStatsInput);
+        boolean anyStatAdded = processRequestedStats(request, stats.get(), neuralStatsInput);
```
Comment (yizheliu-amazon) [inline src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsAction.java:171]: I think you may simplify code a little bit by combining line 161 to 176  into `void processRequestedStats(RestRequest request, NeuralStatsInput neuralStatsInput)`. 

Basically, you can change `void processRequestedStats(RestRequest request, NeuralStatsInput neuralStatsInput)` to do below steps:
1. get `Optional<String[]> stats` from request
2. 
    - 2a: if stats is not present, add all stats to neuralStatsInput and return
    - 2b: if stats is present, continue
3. Validate stats and add stats to `neuralStatsInput` as same as current logic
    -3a: Throw exception if invalidStatNames is not empty
    -3b: Call if necessary addAllStats(neuralStatsInput);
6.

Comment (q-andy) [reply src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsAction.java:171]: Makes sense, I think we can skip the 3b `addAllStats` call as well. When we start validating, we will either be adding specific requested stats (meaning we do not want to add all) or we will be throwing an exception because there are invalid stats (meaning we do not want to add all).

File: src/test/java/org/opensearch/neuralsearch/rest/RestNeuralStatsActionIT.java [hunk 2]
```diff
@@ -120,6 +120,12 @@ public void test_statsFiltering() throws Exception {
         assertNotNull(getNestedValue(stats, InfoStatName.TEXT_EMBEDDING_PROCESSORS.getFullPath()));
     }
 
+    public void test_statsFiltering_invalidThrows() throws Exception {
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/rest/RestNeuralStatsActionIT.java:123]: can you please change approach with testing from integ test to unit tests? We have way too many integ tests, and they are expensive. For such edge case and failure scenario unit test should do the job.

Comment (q-andy) [reply src/test/java/org/opensearch/neuralsearch/rest/RestNeuralStatsActionIT.java]: Sure. Currently text embedding executions is the only event stat. As I add more stats I can refactor the integ tests to not use text embedding models as well so we can avoid model loading too.

File: src/test/java/org/opensearch/neuralsearch/rest/RestNeuralStatsActionTests.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class RestNeuralStatsActionTests extends InferenceProcessorTestCase {
+    private RestNeuralStatsAction restNeuralStatsAction;
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/rest/RestNeuralStatsActionTests.java:45]: why this is class level variable? if it's specific only for test method then it should be scoped at method level

File: src/test/java/org/opensearch/neuralsearch/rest/RestNeuralStatsActionTests.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    NodeClient client;
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/rest/RestNeuralStatsActionTests.java:47]: nit: this can be private

File: src/test/java/org/opensearch/neuralsearch/rest/RestNeuralStatsActionTests.java (hunk 3 above)
Comment (q-andy) [reply src/test/java/org/opensearch/neuralsearch/rest/RestNeuralStatsActionTests.java:45]: Sure I can remove, other tests have them at class level so they can factor out common initialization or cleanup in the setUp and tearDown methods. In this case, different some tests require different setup and there's no special cleanup so it's not necessary.

File: src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsAction.java [hunk 5]
```diff
@@ -157,47 +158,46 @@ private NeuralStatsInput createNeuralStatsInputFromRequestParams(RestRequest req
...
-            return neuralStatsInput;
-        }
+        return neuralStatsInput;
+    }
 
-        // Process requested stats
-        boolean anyStatAdded = processRequestedStats(stats.get(), neuralStatsInput);
+    private void processStatsRequestParameters(RestRequest request, NeuralStatsInput neuralStatsInput) {
+        // Determine which stat names to retrieve based on user parameters
+        Optional<String[]> optionalStats = splitCommaSeparatedParam(request, "stat");
```
[the remark this answers is not in the document]
Comment (junqiu-lei) [reply src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsAction.java:169]: I see it being used in multiple places

**Comment by junqiu-lei** [comment]
Could you please add the updated response example in PR description?

File: src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsAction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (q-andy) [reply src/main/java/org/opensearch/neuralsearch/rest/RestNeuralStatsAction.java]: sure, I factored out `nodeIds` as well

**Comment by q-andy** [comment]
@junqiu-lei added the example response, since we use `unrecognized()` from core. it auto suggests actual stat names based on the invalid stat name which is a nice benefit.

`request [/_plugins/_neural/stats/text_embedding] contains unrecognized stat: [text_embedding] -> did you mean [text_embedding_executions]?`

**Review by junqiu-lei — APPROVED** [review]
LGTM.

**Review by martin-gaievski — APPROVED** [review]
